src/main/java/net/minecraftforge/common/capabilities/Capability.java

```diff
@@ -5,71 +5,345 @@
 
 package net.minecraftforge.common.capabilities;
 
-import net.minecraftforge.common.util.LazyOptional;
-import org.jetbrains.annotations.NotNull;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.function.Consumer;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import org.apache.commons.lang3.mutable.Mutable;
+import org.apache.commons.lang3.mutable.MutableObject;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.MethodsReturnNonnullByDefault;
+import net.minecraft.core.Direction;
+import net.minecraftforge.common.util.NonNullConsumer;
+import net.minecraftforge.common.util.NonNullFunction;
+import net.minecraftforge.common.util.NonNullPredicate;
+import net.minecraftforge.common.util.NonNullSupplier;
 
 /**
- * This is the core holder object Capabilities.
- * Each capability will have ONE instance of this class,
- * and it will the the one passed into the ICapabilityProvider functions.
+ * This object holds a lazily-resolved reference to an instance of a capability
+ * class.  The passed supplier will be executed once needed, and cached.<br>
+ * <br>
+ * Typical transformation operations such as {@link #map(NonNullFunction)} and
+ * {@link #ifPresent(NonNullConsumer)} are available.<br>
+ * <br>
+ * This class also the ability to listen for invalidation, via
+ * {@link #addListener(NonNullConsumer)}. This method is invoked when this
+ * Capability is invalidated via {@link #invalidate()}<br>
+ * <br>
+ * To create an instance of this class, use {@link #of(NonNullSupplier)}.
+ * The result of the supplier must never be null.
+ * <br>
+ * The empty instance can be retrieved with {@link #empty()}.
  *
- * The CapabilityManager is in charge of creating this class.
+ * @param <T> The type of the optional value.
+ * 
+ * @see {@link ICapabilityProvider#getCapability(CapabilityType, Direction)} for retrieving capabilities.
  */
-public class Capability<T>
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+public final class Capability<T>
```

Comment:
Not sure about the renames, they introduce additional migration difficulty with little benefit. Maybe just `LazyOptional` to `CapabilityInstance`. Not very important though, I don't really care.

Reply:
It's a bit of port-work but it shouldn't be so bad, and it is much nicer for readability.  Still not certain if Capability (old LazyOptional) should be Capability or CapabilityInstance / CapabilityReference, kinda waiting on more opinions for that.

Reply:
I would say it's worth it to shift some burden on porting to have a better classname in the long run.